File: lc_soln/0042_trapping_rain_water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        if (n < 3) {
            return 0;
        }

        int l = 0;
        int r = n - 1;
        int lh = height[l];
        int rh = height[r];
        int result = 0;
        int curr;
        while (l < r) {
            if (lh < rh) {
                ++l;
                curr = height[l];
                lh = std::max(lh, curr);
            } else {
                --r;
                curr = height[r];
                rh = std::max(rh, curr);
            }
            result += max(0, min(lh, rh) - curr);
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0042_trapping_rain_water/solution.rs
struct Solution;

impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let n = height.len();
        if n < 3 {
            return 0;
        }
        let mut l = 0;
        let mut r = n - 1;
        let mut lh = height[l];
        let mut rh = height[r];
        let mut result = 0;
        while l < r {
            if lh < rh {
                l += 1;
                lh = lh.max(height[l]);
                result += lh - height[l];
            } else {
                r -= 1;
                rh = rh.max(height[r]);
                result += rh - height[r];
            }
        }
        result
    }
}

fn main() {}
